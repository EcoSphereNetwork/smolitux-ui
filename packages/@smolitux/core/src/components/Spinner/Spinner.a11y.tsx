// packages/@smolitux/core/src/components/Spinner/Spinner.a11y.tsx
import React, { forwardRef, useId } from 'react';
import './Spinner.css';

export interface SpinnerProps extends React.HTMLAttributes<HTMLDivElement> {
  /** Größe des Spinners */
  size?: 'xs' | 'sm' | 'md' | 'lg' | 'xl';
  /** Variante des Spinners */
  variant?: 'border' | 'grow' | 'dots' | 'ring';
  /** Farbe des Spinners */
  color?: 'primary' | 'secondary' | 'success' | 'danger' | 'warning' | 'info' | 'light' | 'dark';
  /** Geschwindigkeit der Animation (in Sekunden) */
  speed?: number;
  /** Label für Screenreader */
  label?: string;
  /** Zentriert den Spinner im Container */
  centered?: boolean;
  /** Zeigt einen Text unter dem Spinner an */
  text?: string;
  /** Volle Breite des Containers */
  fullWidth?: boolean;
  /** ARIA-Label für den Spinner */
  ariaLabel?: string;
  /** Beschreibung für den Spinner (für Screenreader) */
  description?: string;
  /** Ob der Spinner als "busy" markiert werden soll */
  busy?: boolean;
  /** Ob der Spinner als "polite" oder "assertive" angekündigt werden soll */
  liveRegionPoliteness?: 'polite' | 'assertive' | 'off';
  /** Ob der Spinner als "atomic" angekündigt werden soll */
  atomic?: boolean;
  /** Ob der Spinner als "relevant" angekündigt werden soll */
  relevant?: 'additions' | 'removals' | 'text' | 'all';
}

/**
 * Barrierefreie Spinner-Komponente für Ladezustände
 * 
 * @example
 * ```tsx
 * <SpinnerA11y 
 *   size="md" 
 *   variant="border" 
 *   color="primary" 
 *   ariaLabel="Daten werden geladen" 
 * />
 * ```
 */
export const SpinnerA11y = forwardRef<HTMLDivElement, SpinnerProps>(({
  size = 'md',
  variant = 'border',
  color = 'primary',
  speed,
  label = 'Wird geladen...',
  centered = false,
  text,
  fullWidth = false,
  className = '',
  style,
  ariaLabel,
  description,
  busy = true,
  liveRegionPoliteness = 'polite',
  atomic = true,
  relevant,
  ...rest
}, ref) => {
  // Generiere eindeutige IDs für ARIA-Attribute
  const uniqueId = useId();
  const spinnerId = rest.id || `spinner-${uniqueId}`;
  const labelId = `label-${spinnerId}`;
  const descriptionId = description ? `description-${spinnerId}` : undefined;
  const textId = text ? `text-${spinnerId}` : undefined;

  // Größen-spezifische Klassen
  const sizeClasses = {
    xs: 'smolitux-spinner--xs',
    sm: 'smolitux-spinner--sm',
    md: 'smolitux-spinner--md',
    lg: 'smolitux-spinner--lg',
    xl: 'smolitux-spinner--xl'
  };

  // Varianten-spezifische Klassen
  const variantClasses = {
    border: 'smolitux-spinner--border',
    grow: 'smolitux-spinner--grow',
    dots: 'smolitux-spinner--dots',
    ring: 'smolitux-spinner--ring'
  };

  // Farb-spezifische Klassen
  const colorClasses = {
    primary: 'smolitux-spinner--primary',
    secondary: 'smolitux-spinner--secondary',
    success: 'smolitux-spinner--success',
    danger: 'smolitux-spinner--danger',
    warning: 'smolitux-spinner--warning',
    info: 'smolitux-spinner--info',
    light: 'smolitux-spinner--light',
    dark: 'smolitux-spinner--dark'
  };

  // Benutzerdefinierte Animation-Geschwindigkeit
  const customStyle = {
    ...style,
    ...(speed ? { '--spinner-animation-duration': `${speed}s` } as React.CSSProperties : {})
  };

  // Container-Klassen
  const containerClasses = [
    'smolitux-spinner-container',
    centered && 'smolitux-spinner-container--centered',
    fullWidth && 'smolitux-spinner-container--full-width',
    className
  ].filter(Boolean).join(' ');

  // Spinner-Klassen
  const spinnerClasses = [
    'smolitux-spinner',
    sizeClasses[size],
    variantClasses[variant],
    colorClasses[color]
  ].filter(Boolean).join(' ');

  // Render-Logik für verschiedene Varianten
  const renderSpinner = () => {
    switch (variant) {
      case 'dots':
        return (
          <div className={spinnerClasses} style={customStyle} aria-hidden="true">
            <div className="smolitux-spinner-dot"></div>
            <div className="smolitux-spinner-dot"></div>
            <div className="smolitux-spinner-dot"></div>
          </div>
        );
      case 'ring':
        return (
          <div className={spinnerClasses} style={customStyle} aria-hidden="true">
            <div className="smolitux-spinner-ring"></div>
          </div>
        );
      default:
        return (
          <div className={spinnerClasses} style={customStyle} aria-hidden="true"></div>
        );
    }
  };

  // Rendere die versteckte Beschreibung
  const renderDescription = () => {
    if (!description) return null;

    return (
      <div id={descriptionId} className="sr-only">
        {description}
      </div>
    );
  };

  return (
    <>
      {renderDescription()}
      <div 
        ref={ref} 
        id={spinnerId}
        className={containerClasses} 
        role="status"
        aria-label={ariaLabel || label}
        aria-busy={busy}
        aria-live={liveRegionPoliteness}
        aria-atomic={atomic}
        aria-relevant={relevant}
        aria-describedby={descriptionId}
        {...rest}
      >
        {renderSpinner()}
        
        {text && (
          <div 
            id={textId}
            className="smolitux-spinner-text"
          >
            {text}
          </div>
        )}
        
        <span 
          id={labelId}
          className="smolitux-spinner-sr-only"
        >
          {label}
        </span>
      </div>
    </>
  );
});

SpinnerA11y.displayName = 'SpinnerA11y';

export default SpinnerA11y;