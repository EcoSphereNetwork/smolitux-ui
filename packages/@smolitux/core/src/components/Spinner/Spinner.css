/* packages/@smolitux/core/src/components/Spinner/Spinner.css */

/* Container */
.smolitux-spinner-container {
  display: inline-flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
}

.smolitux-spinner-container--centered {
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
}

.smolitux-spinner-container--full-width {
  width: 100%;
}

/* Spinner Base */
.smolitux-spinner {
  display: inline-block;
  vertical-align: middle;
  --spinner-animation-duration: 0.75s;
}

/* Spinner Sizes */
.smolitux-spinner--xs {
  width: 1rem;
  height: 1rem;
}

.smolitux-spinner--sm {
  width: 1.5rem;
  height: 1.5rem;
}

.smolitux-spinner--md {
  width: 2rem;
  height: 2rem;
}

.smolitux-spinner--lg {
  width: 2.5rem;
  height: 2.5rem;
}

.smolitux-spinner--xl {
  width: 3rem;
  height: 3rem;
}

/* Spinner Colors */
.smolitux-spinner--primary {
  --spinner-color: var(--spinner-primary-color, #3b82f6);
}

.smolitux-spinner--secondary {
  --spinner-color: var(--spinner-secondary-color, #6b7280);
}

.smolitux-spinner--success {
  --spinner-color: var(--spinner-success-color, #10b981);
}

.smolitux-spinner--danger {
  --spinner-color: var(--spinner-danger-color, #ef4444);
}

.smolitux-spinner--warning {
  --spinner-color: var(--spinner-warning-color, #f59e0b);
}

.smolitux-spinner--info {
  --spinner-color: var(--spinner-info-color, #3b82f6);
}

.smolitux-spinner--light {
  --spinner-color: var(--spinner-light-color, #f3f4f6);
}

.smolitux-spinner--dark {
  --spinner-color: var(--spinner-dark-color, #1f2937);
}

/* Border Spinner */
.smolitux-spinner--border {
  border-radius: 50%;
  border: 0.2em solid var(--spinner-color);
  border-right-color: transparent;
  animation: smolitux-spinner-border var(--spinner-animation-duration) linear infinite;
}

@keyframes smolitux-spinner-border {
  to {
    transform: rotate(360deg);
  }
}

/* Grow Spinner */
.smolitux-spinner--grow {
  background-color: var(--spinner-color);
  border-radius: 50%;
  animation: smolitux-spinner-grow var(--spinner-animation-duration) ease-in-out infinite;
}

@keyframes smolitux-spinner-grow {
  0% {
    transform: scale(0);
    opacity: 0;
  }
  50% {
    opacity: 1;
  }
  100% {
    transform: scale(1);
    opacity: 0;
  }
}

/* Dots Spinner */
.smolitux-spinner--dots {
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 0.3em;
}

.smolitux-spinner-dot {
  width: 25%;
  height: 25%;
  background-color: var(--spinner-color);
  border-radius: 50%;
  animation: smolitux-spinner-dots var(--spinner-animation-duration) infinite ease-in-out;
}

.smolitux-spinner-dot:nth-child(1) {
  animation-delay: -0.32s;
}

.smolitux-spinner-dot:nth-child(2) {
  animation-delay: -0.16s;
}

@keyframes smolitux-spinner-dots {
  0%, 80%, 100% {
    transform: scale(0);
  }
  40% {
    transform: scale(1);
  }
}

/* Ring Spinner */
.smolitux-spinner--ring {
  position: relative;
}

.smolitux-spinner-ring {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  border: 0.2em solid transparent;
  border-top-color: var(--spinner-color);
  border-radius: 50%;
  animation: smolitux-spinner-ring var(--spinner-animation-duration) linear infinite;
}

.smolitux-spinner-ring::after {
  content: '';
  position: absolute;
  top: 0.15em;
  left: 0.15em;
  right: 0.15em;
  bottom: 0.15em;
  border: 0.2em solid transparent;
  border-top-color: var(--spinner-color);
  border-radius: 50%;
  opacity: 0.7;
  animation: smolitux-spinner-ring-inner calc(var(--spinner-animation-duration) * 0.8) linear infinite;
}

@keyframes smolitux-spinner-ring {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}

@keyframes smolitux-spinner-ring-inner {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(-360deg);
  }
}

/* Text */
.smolitux-spinner-text {
  margin-top: 0.75rem;
  font-size: 0.875rem;
  color: var(--spinner-text-color, #6b7280);
}

/* Screen Reader Only */
.smolitux-spinner-sr-only {
  position: absolute;
  width: 1px;
  height: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  white-space: nowrap;
  border-width: 0;
}

/* Dark Mode */
.dark .smolitux-spinner--primary {
  --spinner-color: var(--spinner-primary-color-dark, #60a5fa);
}

.dark .smolitux-spinner--secondary {
  --spinner-color: var(--spinner-secondary-color-dark, #9ca3af);
}

.dark .smolitux-spinner--success {
  --spinner-color: var(--spinner-success-color-dark, #34d399);
}

.dark .smolitux-spinner--danger {
  --spinner-color: var(--spinner-danger-color-dark, #f87171);
}

.dark .smolitux-spinner--warning {
  --spinner-color: var(--spinner-warning-color-dark, #fbbf24);
}

.dark .smolitux-spinner--info {
  --spinner-color: var(--spinner-info-color-dark, #60a5fa);
}

.dark .smolitux-spinner--light {
  --spinner-color: var(--spinner-light-color-dark, #374151);
}

.dark .smolitux-spinner--dark {
  --spinner-color: var(--spinner-dark-color-dark, #f9fafb);
}

.dark .smolitux-spinner-text {
  color: var(--spinner-text-color-dark, #9ca3af);
}