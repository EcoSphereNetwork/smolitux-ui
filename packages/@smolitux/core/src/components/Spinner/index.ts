// packages/@smolitux/core/src/components/Spinner/index.ts
export { default as Spinner } from './Spinner';
export type { SpinnerProps } from './Spinner';