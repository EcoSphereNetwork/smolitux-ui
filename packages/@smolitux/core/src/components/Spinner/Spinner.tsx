// packages/@smolitux/core/src/components/Spinner/Spinner.tsx
import React, { forwardRef } from 'react';
import './Spinner.css';

export interface SpinnerProps extends React.HTMLAttributes<HTMLDivElement> {
  /** Größe des Spinners */
  size?: 'xs' | 'sm' | 'md' | 'lg' | 'xl';
  /** Variante des Spinners */
  variant?: 'border' | 'grow' | 'dots' | 'ring';
  /** Farbe des Spinners */
  color?: 'primary' | 'secondary' | 'success' | 'danger' | 'warning' | 'info' | 'light' | 'dark';
  /** Geschwindigkeit der Animation (in Sekunden) */
  speed?: number;
  /** Label für Screenreader */
  label?: string;
  /** Zentriert den Spinner im Container */
  centered?: boolean;
  /** Zeigt einen Text unter dem Spinner an */
  text?: string;
  /** Volle Breite des Containers */
  fullWidth?: boolean;
}

/**
 * Spinner-Komponente für Ladezustände
 * 
 * @example
 * ```tsx
 * <Spinner size="md" variant="border" color="primary" />
 * ```
 */
export const Spinner = forwardRef<HTMLDivElement, SpinnerProps>(({
  size = 'md',
  variant = 'border',
  color = 'primary',
  speed,
  label = 'Wird geladen...',
  centered = false,
  text,
  fullWidth = false,
  className = '',
  style,
  ...rest
}, ref) => {
  // Größen-spezifische Klassen
  const sizeClasses = {
    xs: 'smolitux-spinner--xs',
    sm: 'smolitux-spinner--sm',
    md: 'smolitux-spinner--md',
    lg: 'smolitux-spinner--lg',
    xl: 'smolitux-spinner--xl'
  };

  // Varianten-spezifische Klassen
  const variantClasses = {
    border: 'smolitux-spinner--border',
    grow: 'smolitux-spinner--grow',
    dots: 'smolitux-spinner--dots',
    ring: 'smolitux-spinner--ring'
  };

  // Farb-spezifische Klassen
  const colorClasses = {
    primary: 'smolitux-spinner--primary',
    secondary: 'smolitux-spinner--secondary',
    success: 'smolitux-spinner--success',
    danger: 'smolitux-spinner--danger',
    warning: 'smolitux-spinner--warning',
    info: 'smolitux-spinner--info',
    light: 'smolitux-spinner--light',
    dark: 'smolitux-spinner--dark'
  };

  // Benutzerdefinierte Animation-Geschwindigkeit
  const customStyle = {
    ...style,
    ...(speed ? { '--spinner-animation-duration': `${speed}s` } as React.CSSProperties : {})
  };

  // Container-Klassen
  const containerClasses = [
    'smolitux-spinner-container',
    centered && 'smolitux-spinner-container--centered',
    fullWidth && 'smolitux-spinner-container--full-width',
    className
  ].filter(Boolean).join(' ');

  // Spinner-Klassen
  const spinnerClasses = [
    'smolitux-spinner',
    sizeClasses[size],
    variantClasses[variant],
    colorClasses[color]
  ].filter(Boolean).join(' ');

  // Render-Logik für verschiedene Varianten
  const renderSpinner = () => {
    switch (variant) {
      case 'dots':
        return (
          <div className={spinnerClasses} style={customStyle} aria-hidden="true">
            <div className="smolitux-spinner-dot"></div>
            <div className="smolitux-spinner-dot"></div>
            <div className="smolitux-spinner-dot"></div>
          </div>
        );
      case 'ring':
        return (
          <div className={spinnerClasses} style={customStyle} aria-hidden="true">
            <div className="smolitux-spinner-ring"></div>
          </div>
        );
      default:
        return (
          <div className={spinnerClasses} style={customStyle} aria-hidden="true"></div>
        );
    }
  };

  return (
    <div ref={ref} className={containerClasses} role="status" {...rest}>
      {renderSpinner()}
      {text && <div className="smolitux-spinner-text">{text}</div>}
      <span className="smolitux-spinner-sr-only">{label}</span>
    </div>
  );
});

Spinner.displayName = 'Spinner';

export default Spinner;