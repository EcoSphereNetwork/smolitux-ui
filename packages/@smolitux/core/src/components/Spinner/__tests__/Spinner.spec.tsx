// packages/@smolitux/core/src/components/Spinner/__tests__/Spinner.spec.tsx
import React from 'react';
import { render } from '@testing-library/react';
import { Spinner } from '../Spinner';

describe('Spinner Snapshot Tests', () => {
  it('matches snapshot with default props', () => {
    const { container } = render(<Spinner />);
    expect(container).toMatchSnapshot();
  });
  
  it('matches snapshot with size="lg"', () => {
    const { container } = render(<Spinner size="lg" />);
    expect(container).toMatchSnapshot();
  });
  
  it('matches snapshot with variant="dots"', () => {
    const { container } = render(<Spinner variant="dots" />);
    expect(container).toMatchSnapshot();
  });
  
  it('matches snapshot with variant="ring"', () => {
    const { container } = render(<Spinner variant="ring" />);
    expect(container).toMatchSnapshot();
  });
  
  it('matches snapshot with color="success"', () => {
    const { container } = render(<Spinner color="success" />);
    expect(container).toMatchSnapshot();
  });
  
  it('matches snapshot with text', () => {
    const { container } = render(<Spinner text="Bitte warten..." />);
    expect(container).toMatchSnapshot();
  });
  
  it('matches snapshot with centered', () => {
    const { container } = render(<Spinner centered />);
    expect(container).toMatchSnapshot();
  });
  
  it('matches snapshot with fullWidth', () => {
    const { container } = render(<Spinner fullWidth />);
    expect(container).toMatchSnapshot();
  });
});