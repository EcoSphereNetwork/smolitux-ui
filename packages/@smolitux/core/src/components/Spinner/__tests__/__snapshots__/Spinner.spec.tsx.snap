// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner Snapshot Tests matches snapshot with centered 1`] = `
<div>
  <div
    class="smolitux-spinner-container smolitux-spinner-container--centered"
    role="status"
  >
    <div
      aria-hidden="true"
      class="smolitux-spinner smolitux-spinner--md smolitux-spinner--border smolitux-spinner--primary"
    />
    <span
      class="smolitux-spinner-sr-only"
    >
      Wird geladen...
    </span>
  </div>
</div>
`;

exports[`Spinner Snapshot Tests matches snapshot with color="success" 1`] = `
<div>
  <div
    class="smolitux-spinner-container"
    role="status"
  >
    <div
      aria-hidden="true"
      class="smolitux-spinner smolitux-spinner--md smolitux-spinner--border smolitux-spinner--success"
    />
    <span
      class="smolitux-spinner-sr-only"
    >
      Wird geladen...
    </span>
  </div>
</div>
`;

exports[`Spinner Snapshot Tests matches snapshot with default props 1`] = `
<div>
  <div
    class="smolitux-spinner-container"
    role="status"
  >
    <div
      aria-hidden="true"
      class="smolitux-spinner smolitux-spinner--md smolitux-spinner--border smolitux-spinner--primary"
    />
    <span
      class="smolitux-spinner-sr-only"
    >
      Wird geladen...
    </span>
  </div>
</div>
`;

exports[`Spinner Snapshot Tests matches snapshot with fullWidth 1`] = `
<div>
  <div
    class="smolitux-spinner-container smolitux-spinner-container--full-width"
    role="status"
  >
    <div
      aria-hidden="true"
      class="smolitux-spinner smolitux-spinner--md smolitux-spinner--border smolitux-spinner--primary"
    />
    <span
      class="smolitux-spinner-sr-only"
    >
      Wird geladen...
    </span>
  </div>
</div>
`;

exports[`Spinner Snapshot Tests matches snapshot with size="lg" 1`] = `
<div>
  <div
    class="smolitux-spinner-container"
    role="status"
  >
    <div
      aria-hidden="true"
      class="smolitux-spinner smolitux-spinner--lg smolitux-spinner--border smolitux-spinner--primary"
    />
    <span
      class="smolitux-spinner-sr-only"
    >
      Wird geladen...
    </span>
  </div>
</div>
`;

exports[`Spinner Snapshot Tests matches snapshot with text 1`] = `
<div>
  <div
    class="smolitux-spinner-container"
    role="status"
  >
    <div
      aria-hidden="true"
      class="smolitux-spinner smolitux-spinner--md smolitux-spinner--border smolitux-spinner--primary"
    />
    <div
      class="smolitux-spinner-text"
    >
      Bitte warten...
    </div>
    <span
      class="smolitux-spinner-sr-only"
    >
      Wird geladen...
    </span>
  </div>
</div>
`;

exports[`Spinner Snapshot Tests matches snapshot with variant="dots" 1`] = `
<div>
  <div
    class="smolitux-spinner-container"
    role="status"
  >
    <div
      aria-hidden="true"
      class="smolitux-spinner smolitux-spinner--md smolitux-spinner--dots smolitux-spinner--primary"
    >
      <div
        class="smolitux-spinner-dot"
      />
      <div
        class="smolitux-spinner-dot"
      />
      <div
        class="smolitux-spinner-dot"
      />
    </div>
    <span
      class="smolitux-spinner-sr-only"
    >
      Wird geladen...
    </span>
  </div>
</div>
`;

exports[`Spinner Snapshot Tests matches snapshot with variant="ring" 1`] = `
<div>
  <div
    class="smolitux-spinner-container"
    role="status"
  >
    <div
      aria-hidden="true"
      class="smolitux-spinner smolitux-spinner--md smolitux-spinner--ring smolitux-spinner--primary"
    >
      <div
        class="smolitux-spinner-ring"
      />
    </div>
    <span
      class="smolitux-spinner-sr-only"
    >
      Wird geladen...
    </span>
  </div>
</div>
`;
